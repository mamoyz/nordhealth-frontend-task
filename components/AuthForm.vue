<template>
	<provet-card>
		<div class="n-padding-l">
			<provet-stack>
				<provet-input data-testid="username" size="l" :label="$t('auth.form.username_label')" expand required hide-required name="username" type="email" placeholder="user@example.com"></provet-input>
				<provet-input data-testid="password" size="l" :label="$t('auth.form.password_label')" expand required hide-required name="password" :type="passwordInputType" placeholder="••••••••">
					<provet-button data-testid="reveal-password-button" size="l" slot="end" aria-describedby="password-tooltip" square @mousedown="revealPassword(true)" @mouseup="revealPassword(false)">
						<provet-icon v-if="isPasswordRevealed" name="interface-edit-off"></provet-icon>
						<provet-icon v-else name="interface-edit-on"></provet-icon>
					</provet-button>
				</provet-input>
				<provet-button data-testid="submit-button" size="l" type="submit" expand variant="primary">
					<span v-if="page === 'signup'">{{ $t("auth.signup.cta_title") }}</span>
					<span v-else>{{ $t("auth.login.cta_title") }}</span>
				</provet-button>

				<provet-checkbox v-if="page === 'signup'" type="checkbox" :hint="$t('auth.form.updates_hint')" :label="$t('auth.form.updates_label')" expand @change="toggleConfirmedUpdates" :checked="confirmedUpdates"></provet-checkbox>
			</provet-stack>
		</div>
	</provet-card>
	<div v-if="$slots.footer" data-testid="footer" class="n-align-center n-color-text-weaker">
		<slot name="footer"></slot>
	</div>
</template>

<script setup lang="ts">
	interface Props {
		page: "login" | "signup";
	}
	const props = defineProps<Props>();
	const isPasswordRevealed = ref(false);
	const confirmedUpdates = ref(true);
	const passwordInputType = computed(() => (isPasswordRevealed.value ? "text" : "password"));
	const toggleConfirmedUpdates = () => (confirmedUpdates.value = !confirmedUpdates.value);
	const revealPassword = (status: boolean) => (isPasswordRevealed.value = status);
</script>

<style scoped></style>
