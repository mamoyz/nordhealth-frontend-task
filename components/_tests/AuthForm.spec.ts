import { describe, beforeEach, it, expect, vi } from "vitest";
import { mountSuspended } from "@nuxt/test-utils/runtime";
import AuthForm from "@/components/AuthForm.vue";

let wrapper: Awaited<ReturnType<typeof mountSuspended>>;

const createTranslationMock = () => {
	const translations: Record<string, string> = {
		"auth.signup.cta_title": "Create an account",
		"auth.login.cta_title": "Log in",
	};
	return vi.fn((key: string) => translations[key] || key);
};

const mockGlobals = {
	mocks: {
		$t: createTranslationMock(),
	},
};

describe("AuthForm Component", () => {
	beforeEach(async () => {
		wrapper = await mountSuspended(AuthForm, {
			props: {
				page: "login",
			},
			slots: {
				footer: "Test Footer Content",
			},
			global: mockGlobals,
		});
	});

	it("Passes the correct props", async () => {
		expect(wrapper.props("page")).toBe("login");
		await wrapper.setProps({ page: "signup" });
		expect(wrapper.props("page")).toBe("signup");
	});

	it("Renders correct slot", async () => {
		expect(wrapper.find('[data-testid="footer"]').text()).toContain("Test Footer Content");
	});

	it("Toggles isPasswordRevealed on mousedown and mouseup", async () => {
		const revealButton = wrapper.find('[data-testid="reveal-password-button"]');
		await revealButton.trigger("mousedown");
		expect(wrapper.vm.isPasswordRevealed.value).toBe(true);
		expect(wrapper.vm.passwordInputType).toBe("text");

		await revealButton.trigger("mouseup");
		expect(wrapper.vm.isPasswordRevealed.value).toBe(false);
		expect(wrapper.vm.passwordInputType).toBe("password");
	});

	it("Toggles confirmedUpdates when checkbox is changed", async () => {
		await wrapper.setProps({ page: "signup" });
		const checkbox = wrapper.find("provet-checkbox");
		expect(wrapper.vm.confirmedUpdates.value).toBe(true);

		await checkbox.trigger("change");
		expect(wrapper.vm.confirmedUpdates.value).toBe(false);

		await checkbox.trigger("change");
		expect(wrapper.vm.confirmedUpdates.value).toBe(true);
	});

	it("Shows signup button text conditionally", async () => {
		// signup page
		await wrapper.setProps({ page: "signup" });
		const button = wrapper.find('[data-testid="submit-button"]');
		expect(button.text()).toContain("Create an account");

		// login page
		await wrapper.setProps({ page: "login" });
		expect(button.text()).toContain("Log in");
	});

	it("should match the snapshot", () => {
		expect(wrapper.html()).toMatchSnapshot();
	});
});
