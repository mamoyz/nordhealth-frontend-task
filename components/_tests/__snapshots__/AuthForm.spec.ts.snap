// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthForm Component > should match the snapshot 1`] = `
"<provet-card>
  <div class="n-padding-l">
    <provet-stack>
      <provet-input data-testid="username" hide-required=""></provet-input>
      <provet-input data-testid="password" hide-required="">
        <provet-button data-testid="reveal-password-button" slot="end" aria-describedby="password-tooltip">
          <provet-icon></provet-icon>
        </provet-button>
      </provet-input>
      <provet-button data-testid="submit-button"><span>Log in</span></provet-button>
      <!--v-if-->
    </provet-stack>
  </div>
</provet-card>
<div data-testid="footer" class="n-align-center n-color-text-weaker">Test Footer Content</div>"
`;
